### ISSUE TITLE
associations method should return fig and ax

### ISSUE THREAD

**Issue by neongreen13** [shakedzy/dython#24]
Thanks for this amazing work. I have used it several times now to explore my data (with citation). Question - how can I save the heat map visualization?
```
nominal.associations(df_cats, theil_u=True, nominal_columns='all', figsize=(12,12))
```
I tried variations like the following:
```
f2,ax = nominal.associations(df_cats, theil_u=True, nominal_columns='all', figsize=(12,12))
f2.savefig('Categorical Corr Heat Map.png')
```
Thanks!

**Comment by shakedzy** [comment]
Hi @neongreen13 , the `associations` doesn't return the `figure` or the `Axe`. That's actually a good idea.

**Comment by Baukebrenninkmeijer** [comment]
If you're interested; I already did this for my own purposes. The associations method is as follows. Let me know if you want a pull request or something:

```python
def associations(dataset: Union[pd.DataFrame, np.ndarray], cat_cols=None, mark_columns=False, theil_u=False, plot=True, return_results=False, **kwargs):
    """
    Adapted from: https://github.com/shakedzy/dython

    Calculate the correlation/strength-of-association of features in data-set with both categorical (eda_tools) and
    continuous features using:

     - Pearson's R for continuous-continuous cases
     - Correlation Ratio for categorical-continuous cases
     - Cramer's V or Theil's U for categorical-categorical cases

    :param dataset: NumPy ndarray / Pandas DataFrame
        The data-set for which the features' correlation is computed
    :param cat_cols: string / list / NumPy ndarray
        Names of columns of the data-set which hold categorical values. Can also be the string 'all' to state that all
        columns are categorical, or None (default) to state none are categorical
    :param mark_columns: Boolean (default: False)
        if True, output's columns' names will have a suffix of '(nom)' or '(con)' based on there type (eda_tools or
        continuous), as provided by cat_cols
    :param theil_u: Boolean (default: False)
        In the case of categorical-categorical feaures, use Theil's U instead of Cramer's V
    :param plot: Boolean (default: True)
        If True, plot a heat-map of the correlation matrix
    :param return_results: Boolean (default: False)
        If True, the function will return a Pandas DataFrame of the computed associations
    :param kwargs:
        Arguments to be passed to used function and methods
    :return: Pandas DataFrame
        A DataFrame of the correlation/strength-of-association between all features
    """
    if plot is False:
        assert kwargs == {}, f'You have some kwargs that are not needed.\nkwargs: {kwargs}'

    dataset = convert(dataset, 'dataframe')
    columns = dataset.columns
    if cat_cols is None:
        cat_cols = list()
    elif cat_cols == 'all':
        cat_cols = columns
    corr = pd.DataFrame(index=columns, columns=columns)
    for i in range(0, len(columns)):
        for j in range(i, len(columns)):
            if i == j:
                corr[columns[i]][columns[j]] = 1.0
            else:
                if columns[i] in cat_cols:
                    if columns[j] in cat_cols:
                        if theil_u:
                            corr[columns[j]][columns[i]] = theils_u(dataset[columns[i]], dataset[columns[j]])
                            corr[columns[i]][columns[j]] = theils_u(dataset[columns[j]], dataset[columns[i]])
                        else:
                            print(dataset[columns[i]])
                            print(dataset[columns[j]])
                            cell = cramers_v(dataset[columns[i]], dataset[columns[j]])
                            corr[columns[i]][columns[j]] = cell
                            corr[columns[j]][columns[i]] = cell
                    else:
                        cell = correlation_ratio(dataset[columns[i]], dataset[columns[j]])
                        corr[columns[i]][columns[j]] = cell
                        corr[columns[j]][columns[i]] = cell
                else:
                    if columns[j] in cat_cols:
                        cell = correlation_ratio(dataset[columns[j]], dataset[columns[i]])
                        corr[columns[i]][columns[j]] = cell
                        corr[columns[j]][columns[i]] = cell
                    else:
                        cell, _ = ss.pearsonr(dataset[columns[i]], dataset[columns[j]])
                        corr[columns[i]][columns[j]] = cell
                        corr[columns[j]][columns[i]] = cell
    corr.fillna(value=np.nan, inplace=True)
    if mark_columns:
        marked_columns = ['{} (nom)'.format(col) if col in cat_cols else '{} (con)'.format(col) for col in
                          columns]
        corr.columns = marked_columns
        corr.index = marked_columns
    if plot:
        if kwargs.get('ax') is None:
            plt.figure(figsize=kwargs.get('figsize', None))
        cmap = sns.diverging_palette(220, 10, as_cmap=True)
        sns.set(style="white")
        sns.heatmap(corr, annot=kwargs.get('annot', True), fmt=kwargs.get('fmt', '.2f'), cmap=cmap, vmax=1, center=0,
                    square=True, linewidths=kwargs.get('linewidths', 0.5), cbar_kws={"shrink": .8}, cbar=kwargs.get('cbar', True), ax=kwargs.get('ax', None))
        if kwargs.get('ax') is None:
            plt.show()
    if return_results:
        return corr

```

**Comment by shakedzy** [comment]
Hi @Baukebrenninkmeijer ! I'm not a matplotlib expert, so maybe I'm missing something here - but this still doesn't return `ax` and `fig`, right? Am I missing something?

**Comment by Baukebrenninkmeijer** [comment]
No, you're correct, I just realized. For now, it only includes passing a ax. One sec.

**Comment by Baukebrenninkmeijer** [comment]
Ok, so I think this is an API decision. You want to return fig/ax when you want to do something with it after calculating associations. This can be done in two main ways. 

Option 1: return fig/ax from calling the function. 
Option 2: pass an existing ax to the associations function. 

You can then manipulate either the returned ax or passed ax. I think they apply to mostly the same situations. In my implementation, I chose the latter. I think supporting both is a bit unnecessary, but both are valid options. Thoughts?

**Comment by shakedzy** [comment]
I agree. Passing in an `ax` is indeed a good option, regardless of the output. As for the output, I was afraid it might get too complex as there's also an option to return the `corr` DataFrame. I guess we can set another flag for `return_fig_ax`, and not allow both this and `return_results`.. what do you think?

**Comment by Baukebrenninkmeijer** [comment]
Hmmm, it is an solution, but like you said, it might get too complicated. Since it is not a plotting library but rather uses optional plotting, I think just having the ability to pass an ax should be enough in essentially all use-cases. We can be a bit opinionated about API design hahaha. 

So to answer @neongreen13 question, in that case you should do the following:
```python
fig, ax = plt.subplots(figsize=(12, 12))
nominal.associations(df_cats, theil_u=True, nominal_columns='all', ax=ax)
fig.savefig('Categorical Corr Heat Map.png')
```

**Comment by shakedzy** [comment]
Yes, that's seems to be it - we only need to add an `ax=ax` to the `heatmap` function. I think we can make this more "official" by adding an `ax=None` to the `associations` function. I'll give you the honor of opening the pull request, if you'd like :) Thanks!

**Comment by shakedzy** [comment]
merged :) bumped version to 0.3.1
